File: app/src/main/java/com/quran/labs/androidquran/model/bookmark/BookmarkModel.java
package com.quran.labs.androidquran.model.bookmark;

import com.quran.data.model.bookmark.Tag;
import com.quran.data.model.bookmark.Bookmark;
import com.quran.data.model.bookmark.BookmarkData;
import com.quran.labs.androidquran.database.BookmarksDBAdapter;
import com.quran.labs.androidquran.ui.helpers.QuranRow;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.core.util.Pair;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.PublishSubject;
import io.reactivex.rxjava3.subjects.Subject;


@Singleton
public class BookmarkModel {
  private final RecentPageModel recentPageModel;
  private final BookmarksDBAdapter bookmarksDBAdapter;
  private final Subject<Boolean> tagPublishSubject;
  private final Subject<Boolean> bookmarksPublishSubject;

  @Inject
  public BookmarkModel(BookmarksDBAdapter bookmarksAdapter, RecentPageModel recentPageModel) {
    this.recentPageModel = recentPageModel;
    this.bookmarksDBAdapter = bookmarksAdapter;

    tagPublishSubject = PublishSubject.<Boolean>create().toSerialized();
    bookmarksPublishSubject = PublishSubject.<Boolean>create().toSerialized();
  }

  public Observable<Boolean> tagsObservable() {
    return tagPublishSubject.hide();
  }

  public Observable<Boolean> recentPagesUpdatedObservable() {
    return recentPageModel.getRecentPagesUpdatedObservable();
  }

  public Observable<Boolean> bookmarksObservable() {
    return bookmarksPublishSubject.hide();
  }

  public void notifyBookmarksUpdated() {
    bookmarksPublishSubject.onNext(true);
  }

  public void notifyRecentPagesUpdated(int page) {
    recentPageModel.notifyRecentPagesUpdated();
    recentPageModel.updateLatestPage(page);
  }

  public Single<BookmarkData> getBookmarkDataObservable(final int sortOrder) {
    return Single.zip(getTagsObservable(),
        getBookmarksObservable(sortOrder),
        recentPageModel.getRecentPagesObservable(),
        BookmarkData::new)
        .subscribeOn(Schedulers.io());
  }

  public Completable removeItemsObservable(
      final List<QuranRow> itemsToRemoveRef) {
    return Completable.fromCallable(() -> {
      List<Long> tagsToDelete = new ArrayList<>();
      List<Long> bookmarksToDelete = new ArrayList<>();
      List<Pair<Long, Long>> untag = new ArrayList<>();
      for (int i = 0, size = itemsToRemoveRef.size(); i < size; i++) {
        QuranRow row = itemsToRemoveRef.get(i);
        if (row.isBookmarkHeader() && row.tagId > 0) {
          tagsToDelete.add(row.tagId);
        } else if (row.isBookmark() && row.bookmarkId > 0) {
          if (row.tagId > 0) {
            untag.add(new Pair<>(row.bookmarkId, row.tagId));
          } else {
            bookmarksToDelete.add(row.bookmarkId);
          }
        }
      }
      bookmarksDBAdapter.bulkDelete(tagsToDelete, bookmarksToDelete, untag);
      bookmarksPublishSubject.onNext(true);
      if (tagsToDelete.size() > 0) {
        tagPublishSubject.onNext(true);
      }
      return null;
    }).subscribeOn(Schedulers.io());
  }

  public Observable<Long> addTagObservable(final String title) {
    return Observable.fromCallable(() -> {
      long result = bookmarksDBAdapter.addTag(title);
      tagPublishSubject.onNext(true);
      return result;
    }).subscribeOn(Schedulers.io());
  }

  public Completable updateTag(final Tag tag) {
    return Completable.fromCallable(() -> {
      boolean result = bookmarksDBAdapter.updateTag(tag.getId(), tag.getName());
      if (result) {
        tagPublishSubject.onNext(true);
      }
      return null;
    }).subscribeOn(Schedulers.io());
  }

  public Observable<Boolean> updateBookmarkTags(final long[] bookmarkIds,
                                                final Set<Long> tagIds,
                                                final boolean deleteNonTagged) {
    return Observable.fromCallable(() -> {
      boolean result = bookmarksDBAdapter.tagBookmarks(bookmarkIds, tagIds, deleteNonTagged);
      if (result) {
        bookmarksPublishSubject.onNext(true);
      }
      return result;
    }).subscribeOn(Schedulers.io());
  }

  public Observable<Long> safeAddBookmark(final Integer sura, final Integer ayah, final int page) {
    return Observable.fromCallable(() -> {
      long result = bookmarksDBAdapter.addBookmarkIfNotExists(sura, ayah, page);
      bookmarksPublishSubject.onNext(true);
      return result;
    }).subscribeOn(Schedulers.io());
  }

  public Single<List<Tag>> getTagsObservable() {
    return Single.fromCallable(bookmarksDBAdapter::getTags).subscribeOn(Schedulers.io());
  }

  private Single<List<Bookmark>> getBookmarksObservable(final int sortOrder) {
    return Single.fromCallable(() -> bookmarksDBAdapter.getBookmarks(sortOrder));
  }

  public Maybe<List<Long>> getBookmarkTagIds(Single<Long> bookmarkIdSingle) {
    return bookmarkIdSingle.filter(bookmarkId -> bookmarkId > 0)
        .map(bookmarksDBAdapter::getBookmarkTagIds)
        .subscribeOn(Schedulers.io());
  }

  public Single<Long> getBookmarkId(final Integer sura, final Integer ayah, final int page) {
    return Single.fromCallable(() -> bookmarksDBAdapter.getBookmarkId(sura, ayah, page))
        .subscribeOn(Schedulers.io());
  }

  public Observable<List<Bookmark>> getBookmarkedAyahsOnPageObservable(Integer... pages) {
    return Observable.fromArray(pages)
        .map(bookmarksDBAdapter::getBookmarkedAyahsOnPage)
        .filter(bookmarks -> !bookmarks.isEmpty())
        .subscribeOn(Schedulers.io());
  }

  public Observable<Pair<Integer, Boolean>> getIsBookmarkedObservable(Integer... pages) {
    return Observable.fromArray(pages)
        .map(page -> new Pair<>(page, bookmarksDBAdapter.getBookmarkId(null, null, page) > 0))
        .subscribeOn(Schedulers.io());
  }

  public Single<Boolean> toggleBookmarkObservable(
      final Integer sura, final Integer ayah, final int page) {
    return getBookmarkId(sura, ayah, page)
        .map(bookmarkId -> {
          boolean result;
          if (bookmarkId > 0) {
            bookmarksDBAdapter.removeBookmark(bookmarkId);
            result = false;
          } else {
            bookmarksDBAdapter.addBookmark(sura, ayah, page);
            result = true;
          }
          bookmarksPublishSubject.onNext(true);
          return result;
        }).subscribeOn(Schedulers.io());
  }

  public Observable<Boolean> importBookmarksObservable(final BookmarkData data) {
    return Observable.fromCallable(() -> {
      boolean result = bookmarksDBAdapter.importBookmarks(data);
      if (result) {
        bookmarksPublishSubject.onNext(true);
      }
      return result;
    }).subscribeOn(Schedulers.io()).cache();
  }
}

File: app/src/main/java/com/quran/labs/androidquran/model/bookmark/BookmarkModel.kt
package com.quran.labs.androidquran.model.bookmark

import androidx.core.util.Pair
import com.quran.data.di.AppScope
import com.quran.data.model.bookmark.Bookmark
import com.quran.data.model.bookmark.BookmarkData
import com.quran.data.model.bookmark.Tag
import com.quran.labs.androidquran.database.BookmarksDBAdapter
import com.quran.labs.androidquran.ui.helpers.QuranRow
import com.quran.mobile.bookmark.model.isDefaultBookmarkCollectionId
import dev.zacsweers.metro.Inject
import dev.zacsweers.metro.SingleIn
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import io.reactivex.rxjava3.subjects.PublishSubject

@Deprecated("") // Use BookmarksDao instead
@SingleIn(AppScope::class)
open class BookmarkModel @Inject constructor(
  private val bookmarksDBAdapter: BookmarksDBAdapter
) {
  private val tagPublishSubject =
    PublishSubject.create<Boolean>().toSerialized()
  private val bookmarksPublishSubject =
    PublishSubject.create<Boolean>().toSerialized()

  open fun tagsObservable(): Observable<Boolean> {
    return tagPublishSubject.hide()
  }

  fun bookmarksObservable(): Observable<Boolean> {
    return bookmarksPublishSubject.hide()
  }

  fun notifyBookmarksUpdated() {
    bookmarksPublishSubject.onNext(true)
  }

  open fun getBookmarkDataObservable(sortOrder: Int): Single<BookmarkData> {
    return Single.zip(
      tagsObservable,
      getBookmarksObservable(sortOrder)
    ) { tags: List<Tag>, bookmarks: List<Bookmark> ->
      BookmarkData(
        tags,
        bookmarks
      )
    }
      .subscribeOn(Schedulers.io())
  }

  fun removeItemsObservable(
    itemsToRemoveRef: List<QuranRow>
  ): Completable {
    return Completable.fromCallable {
      val tagsToDelete: MutableList<Long> =
        ArrayList()
      val bookmarksToDelete: MutableList<Long> =
        ArrayList()
      val untag: MutableList<Pair<Long, Long>> =
        ArrayList()
      var i = 0
      val size = itemsToRemoveRef.size
      while (i < size) {
        val row = itemsToRemoveRef[i]
        val tagId = row.oldDatabaseTagId()
        val bookmarkId = row.bookmarkId?.oldDatabaseIdToLongOrNull()
        if (row.isBookmarkHeader && tagId != null) {
          tagsToDelete.add(tagId)
        } else if (row.isBookmark && bookmarkId != null) {
          if (tagId != null) {
            untag.add(
              Pair(
                bookmarkId,
                tagId
              )
            )
          } else {
            bookmarksToDelete.add(bookmarkId)
          }
        }
        i++
      }
      bookmarksDBAdapter.bulkDelete(tagsToDelete, bookmarksToDelete, untag)
      bookmarksPublishSubject.onNext(true)
      if (tagsToDelete.isNotEmpty()) {
        tagPublishSubject.onNext(true)
      }
      null
    }.subscribeOn(Schedulers.io())
  }

  fun addTagObservable(title: String): Observable<Long> {
    return Observable.fromCallable {
      val result = bookmarksDBAdapter.addTag(title)
      tagPublishSubject.onNext(true)
      result
    }.subscribeOn(Schedulers.io())
  }

  fun updateTag(tag: Tag): Completable {
    return Completable.fromCallable {
      val result = tag.id.oldDatabaseIdToLongOrNull()
        ?.let { tagId -> bookmarksDBAdapter.updateTag(tagId, tag.name) }
        ?: false
      if (result) {
        tagPublishSubject.onNext(true)
      }
      null
    }.subscribeOn(Schedulers.io())
  }

  open fun updateBookmarkTags(
    bookmarkIds: LongArray,
    tagIds: Set<Long>,
    deleteNonTagged: Boolean
  ): Observable<Boolean> {
    return Observable.fromCallable {
      val result =
        bookmarksDBAdapter.tagBookmarks(bookmarkIds, tagIds, deleteNonTagged)
      if (result) {
        bookmarksPublishSubject.onNext(true)
      }
      result
    }.subscribeOn(Schedulers.io())
  }

  open fun safeAddBookmark(sura: Int?, ayah: Int?, page: Int): Observable<Long> {
    return Observable.fromCallable {
      val result = bookmarksDBAdapter.addBookmarkIfNotExists(sura, ayah, page)
      bookmarksPublishSubject.onNext(true)
      result
    }.subscribeOn(Schedulers.io())
  }

  open val tagsObservable: Single<List<Tag>>
    get() = Single.fromCallable<List<Tag>> { bookmarksDBAdapter.getTags() }
      .subscribeOn(Schedulers.io())

  private fun getBookmarksObservable(sortOrder: Int): Single<List<Bookmark>> {
    return Single.fromCallable {
      bookmarksDBAdapter.getBookmarks(
        sortOrder
      )
    }
  }

  open fun getBookmarkTagIds(bookmarkIdSingle: Single<Long>): Maybe<List<Long>> {
    return bookmarkIdSingle.filter { bookmarkId: Long -> bookmarkId > 0 }
      .map { bookmarkId: Long ->
        bookmarksDBAdapter.getBookmarkTagIds(bookmarkId)
      }
      .subscribeOn(Schedulers.io())
  }

  fun getBookmarkId(sura: Int?, ayah: Int?, page: Int): Single<Long> {
    return Single.fromCallable {
      bookmarksDBAdapter.getBookmarkId(
        sura,
        ayah,
        page
      )
    }
      .subscribeOn(Schedulers.io())
  }

  private fun QuranRow.oldDatabaseTagId(): Long? {
    return tagId
      ?.takeUnless { tagId -> tagId.isDefaultBookmarkCollectionId() }
      ?.oldDatabaseIdToLongOrNull()
  }

  private fun String.oldDatabaseIdToLongOrNull(): Long? = toLongOrNull()
}
